lighthouse-core/gather/gatherers/form-elements.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ * @return {LH.Artifacts['FormElements']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+  // @ts-ignore - put into scope via stringification
+  const formChildren = getElementsInDocument('textarea, input, label, select'); // eslint-disable-line no-undef
+  const forms = new Map();
+  /** @type { { inputs: Array<LH.Artifacts.FormInput>, labels: Array<LH.Artifacts.FormLabel> }  } */
+  const formlessObj = {
+    inputs: [],
+    labels: [],
+  };
+  for (const child of formChildren) {
+    const parentFormElement = child.closest('form');
+    const hasForm = !!parentFormElement;
+    if (hasForm && !forms.has(parentFormElement)) {
+      const newFormObj = {
+        attributes: {
+          id: parentFormElement.id,
+          name: parentFormElement.name,
+          autocomplete: parentFormElement.autocomplete,
+          // @ts-ignore - put into scope via stringification
+          nodeLabel: getNodeLabel(child), // eslint-disable-line no-undef,
+          // @ts-ignore - put into scope via stringification
+          snippet: getOuterHTMLSnippet(parentFormElement), // eslint-disable-line no-undef
+        },
+        inputs: [],
+        labels: [],
+      };
+      forms.set(parentFormElement, newFormObj);
+    }
+    const formObj = hasForm ? forms.get(parentFormElement) : formlessObj;
+
+    if (child instanceof HTMLInputElement || child instanceof HTMLTextAreaElement) {
+      const isButton = child instanceof HTMLInputElement &&
+      (child.type === 'submit' || child.type === 'button');
+      if (isButton) continue;
+      formObj.inputs.push({
+        id: child.id,
+        name: child.name,
+        placeholder: child.placeholder,
+        autocomplete: child.autocomplete,
+        // @ts-ignore - put into scope via stringification
+        nodeLabel: getNodeLabel(child), // eslint-disable-line no-undef,
+        // @ts-ignore - put into scope via stringification
+        snippet: getOuterHTMLSnippet(child), // eslint-disable-line no-undef
+      });
+    }
+    if (child instanceof HTMLSelectElement) {
```

Comment:
Why is this a different branch from the above? my isButton suggestion was mostly so these could be combined :)

Reply:
the reason they were seperated wasnt because of the button, but because select elements dont have a placeholder attribute. I can merge it together thought if I use the getAttribute function

Reply:
or `child instanceof HTMLSelectElement ? undefined : child.placeholder` ? just seems a bit much to duplicate the entire block for a single property changing